### PR TITLE
Several fixes and changes to the dispatchers

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change ###

This PR makes a few fixes to the #2862 PR in addition to the #3436 fix:

 - iOS was using `NSRunLoop` which is not correct. Now using `DispatcherQueue` 
 - Android always compared the dispatcher looper (which is always the `MainLooper`) to the `MainLooper` which resulted in `IsDispatchRequired` always returning false - even on background threads
 - Renamed the incorrect name of the `BeginInvokeOnMainThread` method (and related) in Core:
    - `IDispatcher.IsDispatchRequired` property
    - `bool IDispatcher.Dispatch` method (now also returns `bool` if the dispatch was added to the queue successfully
    - `IDispatcher.DispatchAsync` extension method
 - The various methods on `Device` in controls is still the same for backwards compatibility

### CODE REVIEW

File: src/Controls/src/Core/DispatcherExtensions.cs [hunk 1]
```diff
@@ -28,28 +29,62 @@ public static IDispatcher GetDispatcher(this BindableObject? bindableObject)
...
+		{
+			dispatcher = EnsureDispatcher(dispatcher);
+			if (dispatcher.IsDispatchRequired)
 			{
-				dispatcher.BeginInvokeOnMainThread(action);
+				return dispatcher.DispatchAsync(action);
 			}
 			else
 			{
 				action();
```
Comment (AmrAlSayed0) [inline src/Controls/src/Core/DispatcherExtensions.cs:68]: `return action();` ? The task returned is not being awaited in this case. It is being awaited inside the [DispatchAsync](https://github.com/dotnet/maui/blob/94781e8a70bee327b76edb179d032d006b8327ea/src/Core/src/Dispatching/DispatcherExtensions.cs#L9-L27) though.

Comment (mattleibow) [reply src/Controls/src/Core/DispatcherExtensions.cs:68]: Good catch!

File: src/Controls/samples/Controls.Sample/Startup.cs [hunk 2]
```diff
@@ -209,7 +209,7 @@ public static MauiApp CreateMauiApp()
 #elif WINDOWS
 					// Log everything in this one
 					events.AddWindows(windows => windows
-						.OnNativeMessage((a, b) => LogEvent(nameof(WindowsLifecycle.OnNativeMessage)))
+						//.OnNativeMessage((a, b) => LogEvent(nameof(WindowsLifecycle.OnNativeMessage)))
```
Comment (PureWeen) [inline src/Controls/samples/Controls.Sample/Startup.cs:212]: Did you mean to leave this commented out?

Comment (mattleibow) [reply src/Controls/samples/Controls.Sample/Startup.cs:212]: yep, that does a spam of logs to the console and makes windows run slow. Dragging the window is like 100k messages.

File: src/Controls/src/Core/Items/MarshalingObservableCollection.cs [hunk 3]
```diff
@@ -59,7 +59,7 @@ void InternalCollectionChanged(object sender, NotifyCollectionChangedEventArgs a
 				args = new ResetNotifyCollectionChangedEventArgs(items);
 			}
 
-			_dispatcher.Dispatch(() => HandleCollectionChange(args));
+			_dispatcher.DispatchIfRequired(() => HandleCollectionChange(args));
```
Comment (PureWeen) [inline src/Controls/src/Core/Items/MarshalingObservableCollection.cs:62]: Does this change behavior slightly? Now it'll conditionally dispatch whereas before it always dispatched?

Comment (mattleibow) [reply src/Controls/src/Core/Items/MarshalingObservableCollection.cs:62]: I just renamed the extension method. A bit confusing that before it was Invoke and the extension Dispatch would do the checks first. Now the method is Dispatch and the extension is DispatchIfRequred.

Before and after is checking.

**Review by PureWeen — APPROVED** [review]

**Pull request merged**